File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/command/uielement/UiControl.java
package com.bigboxcode.designpattern.command.uielement;

import java.util.ArrayList;
import java.util.List;

public class UiControl {
    private List<UiCommand> commandList = new ArrayList<>();

    public void addElement(UiCommand command) {
        // Execute command
        command.print();

        // Store command in list to have a history
        commandList.add(command);
    }

    public void removeElement(UiCommand command) {
        // Remove element
        command.remove();

        // Store command in list to have a history
        commandList.remove(command);
    }

    public void undo() {
        UiCommand lastCommand = commandList.get(commandList.size() - 1);
        removeElement(lastCommand);
    }
}

File: design-patterns/ts/command/ui-element/ui-control.ts
// ui-control.ts

import UiCommand from "./ui-command";

class UiControl {
    private commandList: UiCommand[] = [];

    public addElement(command: UiCommand) {
        command.print();

        this.commandList.push(command);
    }

    public removeElement(command: UiCommand) {
        command.remove();

        for (let i = 0; i < this.commandList.length; i++) {
            if (command == this.commandList[i]) {
                this.commandList.splice(i, 1);
                break;
            }
        }
    }

    public undo() {
        const lastCommand = this.commandList[this.commandList.length - 1];
        this.removeElement(lastCommand);
    }
}

export default UiControl;